Feat: #2 공통 예외처리 기능 개발

KimJayhyun opened the pull request:
## PR Type

What kind of change does this PR introduce?

<!-- 'x'를 이용하여 이 PR에 적용되는 항목을 확인해 주세요. -->

- [x] **\[Feat\]** 🎨 새로운 기능을 추가했어요.
- [x] **\[Config\]** ⚙ 환경설정을 변경했어요.

## Related Issues

<!--#을 눌러 이슈와 연결해 주세요-->
Close #2 
## What does this PR do?

<!--무엇을 하셨나요?-->

- [x] Custom Exception Class를 개발함
- [x] `GlobalExceptionHandler`를 개발함
- [x] `ErrorDto`를 활용하여 클라이언트에 Response함

## Other information

<!--참고한 자료, 추가적인 사항, 기타 의견-->
- 테스트로 사용했던 `testRestController`를 삭제하고 `assertThrows`를 사용해서 테스트를 했어요!
- 확장에 용이하도록 `BusinessException`에서 `errorDto`를 초기화했어요.
   - 다른 Exception Class를 생성할 때에도  `errorCode`만 생성자로 주입받으면 되요.
   - `SonarLint`의 조언 중 객체의 필드 중 직렬화를 하지 않을 필드를 선택하는 `transient`를 사용하라는 경고가 있었어요.
   - 이 방식의 개발이 좋은지는 같이 얘기해보면 좋겠어요!

KimJayhyun committed:
Feat: #2 공통 예외처리 & 핸들러 기능 구현

패키지 구조 생성

KimJayhyun committed:
Feat: #2 Exception Class 및 전역 Exception Handler 구현

KimJayhyun committed:
Feat: #2 커스텀에러 DTO & `BusinessException` 추가

Commit:
Merge branch 'develop' into feature/#2-공통예외처리&핸들러

thornscript commented on src/main/java/com/growup/pms/common/exception/dto/ErrorDto.java:
```diff
@@ -0,0 +1,19 @@
+import lombok.Setter;
+import org.springframework.http.HttpStatus;
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
## DTO와 롬복
이 부분은 문제되는 부분은 아니지만, 통일성을 위해서 DTO에 대한 컨벤션도 정해보면 어떨까 싶어서 코멘트를 남겨봅니다. [영한님의 글](https://www.inflearn.com/questions/161417/dto-%EC%82%AC%EC%9A%A9%EC%97%90%EB%8C%80%ED%95%B4-%EA%B6%81%EA%B8%88%ED%95%A9%EB%8B%88%EB%8B%A4)에서 아래와 같은 부분을 확인할 수 있습니다!

> Entity는 비즈니스 로직이 있고, 실제 데이터도 변경되기 때문에, Setter를 최대한 사용하지 않는 편이 좋습니다.
> 
> 반면에 DTO는 목적 자체가 어떤 로직이 있다기 보다는, 단순히 데이터를 전달하는 것이기 때문에 Getter, Setter를 자유롭게 사용하셔도 됩니다^^

편의상 강의에서 DTO에 `@Data`를 붙이시기도 하시더라구요! 이 부분에 대해서는 개발자마다 개인적인 차이가 있을 수 있지만, DTO의 데이터는 변경되어도 큰 영향은 없으나 변경될 일이 잘 없으므로 아래와 같이 통일해보는 건 어떨까 싶어서요!

```java
// 클라이언트에게 노출되는 생성자나 메서드를 줄여 혼동을 방지함
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
```

## 클래스 레벨의 @Builder
클래스 레벨에 `@Builder`를 달면 `@AllArgsConstructor(access = AccessLevel.PACKAGE)`를 추가하고 이 모든 필드를 받는 생성자에 `@Builder` 애노테이션을 적용하는 것과 동일하게 됩니다. 이 경우에는 동일 패키지에 있는 다른 클래스에서 생성자에 접근할 수 있게 되며, 일부 필드가 필수적이지 않은 경우 이를 조정하기가 어렵다는 단점이 있습니다! 🐱 

그리고 롬복은 사용자가 직접 생성자를 작성했거나 명시적인 생성자 애노테이션(`@XArgsConstructor`)을 사용한 경우, 롬복은 모든 필드를 받는 생성자가 존재한다고 가정하고 해당 생성자를 사용하려고 시도하므로 개발자의 의도와는 다르게 컴파일러 에러가 발생할 수 있으므로 클래스 레벨에 `@Builder`를 선언할 때는 주의해야 할 것 같아요.

그래서 아래와 같이 일관성을 위해서 아래와 같이 선언해보는 건 어떨까요?
```java
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ErrorDto {
    private HttpStatus httpStatus;
    private String code;
    private String message;

    @Builder
    public ErrorDto(HttpStatus httpStatus, String code, String message) {
        this.httpStatus = httpStatus;
        this.code = code;
        this.message = message;
    }
}
```

KimJayhyun replied:
이번 DTO를 작성하면서 했던 고민을 도현님이 너무 잘 이해하고 계시네요...

`Entity`에는 `Setter` 추가하지 않는 것이 좋다는 가이드를 저도 보았던 기억이 있어요!
허나 `DTO`에는 그런 제약이 없어서 그런지 코드 작성 전에 찾아본 여러 사람들의 `DTO` 작성 방법은 천차 만별이더라구요..!

그래서 `@Data`도 좋지만 명시적인 걸 좋아하는 저는 저런 어노테이션 무덤을 만들어 놨습니다... ㅋㅋㅋ

`DTO`에 대한 컨벤션이 필요하다는 생각을 했었어요..!

```Java
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ErrorDto {
    private HttpStatus httpStatus;
    private String code;
    private String message;

    @Builder
    public ErrorDto(HttpStatus httpStatus, String code, String message) {
        this.httpStatus = httpStatus;
        this.code = code;
        this.message = message;
    }
}
```

말씀하신데로 위와 같이 `DTO`에 대한 컨벤션을 가져가는 것이 좋을거 같아요!

좋아하는 코드 철학 중에 이런 말이 있어요.

아름다운 것이 추한 것보다 낫다.
명시적인 것이 암시적인 것보다 낫다.

들어보셨겠지만 Python 철학에 대한 얘기에요!

한 번 더 읽어보시라구 링크도 남겨드릴게요..!
[파이썬의 선 (The Zen of Python)](https://wikidocs.net/7907)

KimJayhyun replied:
`@Builder`에 대한 공부는 제가 조금 더 해보도록 할게요..!

이론적인걸 보고 개발을 해야했는데 수박 겉 핥기식이네요 ㅠㅠ

화이팅해보겠습니다.

자세한 리뷰 너무 감사해요!

kimpearl3599 replied:
낭만있다.... 크 :)

thornscript replied:
@KimJayhyun [공식 문서](https://projectlombok.org/features/Builder)와 [정리가 잘 된 블로그](https://velog.io/@park2348190/Lombok-Builder%EC%9D%98-%EB%8F%99%EC%9E%91-%EC%9B%90%EB%A6%AC)가 있어서 추천드려요! 슬쩍 제 [블로그 글](https://blog.hexabrain.net/398)도 끼어넣어 봅니다... 🥹 사실 각각의 방법은 모두 장단점이 존재하고 스스로가 옳다고 생각하면 다른 팀원과의 협의 하에 정해도 좋다고 생각합니다! 정답은 없으니까요... 🐈 

아무래도 일정이 있으시다 보니까 그렇기도 하고, 토이 프로젝트를 진행하면서 별도로 독서하는 시간까지 할애하시고 계신데 사실 라이브러리나 프레임워크의 구체적인 동작 원리까지 신경쓰기에는 너무나도 벅찰 것 같아서 이렇게 작성하시는 게 너무나도 당연스럽습니다... 😢 저도 처음에 내부 동작을 모르고 피드백을 받고서 성장한 부분이라 이 부분을 재현님에게도 공유하고 싶었습니다.

thornscript replied:
@KimJayhyun 다른 문구도 깊게 와닿고 반성하게 되는 부분도 있는 것 같습니다. 자바의 철학과도 일부 일맥상통하는 파이썬의 좋은 철학과 정보 공유해주셔서 감사합니다! 특히 '지금 하는 것이 전혀 안하는 것보다 낫다.' 이 부분이 여러 의미로 제게 꽂히네요... 😢 리팩토링의 필요성을 느끼고도 미뤄뒀다가 나중에 비용이 너무 커져서 오히려 방치하게 되고 프로젝트의 전체적인 퀄리티가 떨어지는 문제를 최근에 경험했어서 더욱 그런 것 같습니다.... 🥲

thornscript commented on src/main/java/com/growup/pms/common/exception/dto/ErrorDto.java:
```diff
@@ -0,0 +1,19 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ErrorDto {
```
이건 제 사견이긴 한데, 보통 DTO의 접두사에 `Dto`, `Request`, `Response` 등을 많이 사용하는 편인데 단순히 특정 계층에 종속되지 않는 단순 전달용 객체를 넘어서 클라이언트에게 응답으로 반환되는 객체이므로 요청이나 응답에 대한 DTO 객체의 postfix는 `Request`, `Response`라는 접두사로 통일해보는 건 어떨까 싶어서요! 그 외의 경우에는 `Dto`로 통일해 보는게 어떨까 하는 게 제 생각이었습니다! 🐱

KimJayhyun replied:
`DTO`에 대한 컨벤션을 정확하게 정의를 하고 넘어가는 것이 좋을 것 같네요...!

저는 말씀하시는 방향이 가독성을 높이는 방법인 것 같아요..!

우리 `DTO` 컨벤션에 대해서 확실하게 정하고 이후 개발에 들어가봐요!

thornscript replied:
@KimJayhyun 금요일 날 오후 3시에 엔티티를 같이 설계해보면서 컨벤션도 같이 논의해보면 좋을 것 같습니다! 👍

thornscript commented on src/main/java/com/growup/pms/common/exception/exceptions/AuthenticationException.java:
```diff
@@ -0,0 +1,13 @@
+
+import lombok.Getter;
+
+@Getter
```
롬복의 `@Getter`는 기본 access modifier가 `public`이라서 `BusinessException`을 상속받는 `AuthenticationException`도 BusinessException의 `getErrorCode()`를 물려받게 됩니다! 따라서 아래와 같이 `BusinessException`의 클래스 레벨에서만 @Getter를 선언해도 괜찮을 것 같아요! 🐱 

```java
@Getter
public class BusinessException extends RuntimeException {
    private final ErrorCode errorCode;

    public BusinessException(ErrorCode errorCode) {
        super(errorCode.getMessage());
        this.errorCode = errorCode;
    }
}

public class AuthenticationException extends BusinessException {
    public AuthenticationException(ErrorCode errorCode) {
        super(errorCode);
    }
}
```

Qodana가 지적한 부분도 이 부분인 것 같네요! 🐈

KimJayhyun replied:
코드를 보면서 뭔가 이상한데...?? 🤔 직감적으로 느꼈던 부분이 이 부분이네요.

언넝 성장해서 이런 리뷰의 수를 줄여볼게요..!!

세세한 지적 너무 감사해요

thornscript commented on src/main/java/com/growup/pms/common/exception/exceptions/BusinessException.java:
```diff
@@ -0,0 +1,20 @@
+@Getter
+public class BusinessException extends RuntimeException {
+    private final ErrorCode errorCode;
+    private final transient ErrorDto errorDto;
```
## 객체 생성의 책임
이 부분에 대해서는 `ErrorDto` 내부에 `of`라는 정적 팩토리 메서드를 만들어서 객체 생성의 책임을 `ErrorDto`에 맡기는 것이 좋은 것 같아요! 🐱 

```java
@Getter
public class BusinessException extends RuntimeException {
    private final ErrorCode errorCode;

    public BusinessException(ErrorCode errorCode) {
        super(errorCode.getMessage());
        this.errorCode = errorCode;
    }
}

@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ErrorDto {
    private HttpStatus httpStatus;
    private String code;
    private String message;

    public ErrorDto(HttpStatus httpStatus, String code, String message) {
        this.httpStatus = httpStatus;
        this.code = code;
        this.message = message;
    }

    public static ErrorDto of(ErrorCode code) {
        return new ErrorDto(code.getHttpStatus(), code.getCode(), code.getMessage());
    }
}
```

## 정적 팩토리 메서드에 흔히 사용하는 명명 방식
아마도 [Effective Java](https://product.kyobobook.co.kr/detail/S000001033066)에서 봤던 것 같은데, 보통은 아래와 같은 명명 방식들을 따른다고 해요! 참고로 봐주시면 감사하겠습니다...! 🐈 

- <b>from:</b> 매개변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 형 변환 메서드
- <b>of:</b> 여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드
- <b>valueOf:</b> from과 of의 더 자세한 버전
- <b>instance 혹은 getlnstance:</b> (매개변수를 받는다면) 매개변수로 명시한 인스턴스를 반환하지만, 같은 인스턴스임을 보장하지는 않는다.
- <b>create 혹은 newlnstance:</b> instance 혹은 getlnstance와 같지만, 매번 새로운 인스턴스를 생성해 반환함을 보장한다.
- <b>getType:</b> getlnstance와 같으나, 생성할 클래스가 아닌 다른 클래스에 팩토리 메서드를 정의할 때 쓴다. "Type"은 팩토리 메서드가 반환할 객체의 타입이다.
- <b>newType:</b> newlnstance와 같으나, 생성할 클래스가 아닌 다른 클래스에 팩토리 메서드를 정의할 때 쓴다. "Type"은 팩토리 메서드가 반환할 객체의 타입이다.
- <b>type:</b> getType과 newType의 간결한 버전

```java
Date d = Date.from(instant); // from
Set<Rank> faceCards = EnumSet.of(JACK, QUEEN, KING); // of
BigInteger prime = BigInteger.valueOf(Integer.MAX_VALUE); // valueOf
StackWalker luke = StackWalker.getInstance(options); // getInstance
Object newArray = Array.newInstance(classObject, arrayLen); // create or newInstance
Filestore fs = Files.getFileStore(path); // getType
BufferedReader br = Files.newBufferedReader(path); // newType
List<Complaint> litany = Collections.list(legacyLitany); // type
```

찾아보니까 [여기](https://velog.io/@saint6839/%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%86%A0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C-%EB%84%A4%EC%9D%B4%EB%B0%8D-%EB%B0%A9%EC%8B%9D)에도 동일한 글이 정리되어 있는 것 같아서 참고하시면 좋을 것 같아요! 🐈‍⬛

KimJayhyun replied:
객체 생성의 책임을 객체에 맡기면서 코드의 가독성을 높히고
개발자가 객체를 생성할 때 세세한 정보를 몰라도 되는 장점을 가져갈 수  있군요..!

위 정적 팩토리 메서드에 대해 정리해두고 다음에도 사용해보겠습니다!!

thornscript commented on src/main/java/com/growup/pms/common/exception/dto/ErrorDto.java:
```diff
@@ -0,0 +1,19 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ErrorDto {
+    private HttpStatus httpStatus;
```
클라이언트가 응답으로 받을 때 `HttpStatus`는 이미 HTTP 응답 코드에 포함되어 있어서 아래와 같이 `@JsonIgnore`를 통해서 직렬화 시에 이 필드를 무시할 수 있도록 만드는 것도 좋을 것 같아요...! 🐱 

```java
@JsonIgnore
private HttpStatus httpStatus;
```

KimJayhyun replied:
아 이것때문에 뜬금없는 `transient`에 대한 얘기가 나왔나보군요..!

말씀하신 방법과 한 가지 방법이 더 있을 것 같아요!
1. `@JsonIgnore`을 추가한다.
2. 코드를 아래와 같은 방향으로 리팩토링한다.
- `ErrorCode`에서 `HttpStatus`를 제외한다.
- `GlobalExceptionHandler`에서 `ErrorCode`의 `HttpStatus`가 아닌 개발자가 직접 설정을 한다.

두 방법 모두 장단점이 있는 것 같아요.
어떻게 보면 라이브러리, 프레임워크가 어디까지 개발자에게 자유를 줄 것인가 같은 고민이 될 수도 있겠네요.

개인적으로는 `@JsonIgnore`을 추가하는게 좋은 것 같아요!
(저같이 하지말라는 걸 꼭 해보는 개발자들이 많거든요 ㅋㅋㅋ)

도현님의 생각은 어떠실까요..?

thornscript replied:
> - `ErrorCode`에서 `HttpStatus`를 제외한다.
> - `GlobalExceptionHandler`에서 `ErrorCode`의 `HttpStatus`가 아닌 개발자가 직접 설정을 한다.

이 경우에는 그런 장점이 있을 것 같아요! 같은 카테고리 내의 에러들은 보통 같은 상태 코드(예: 인가 AUTHZ와 관련해서는 403, 인증 AUTH과 관련해서는 401 등)를 가지고 있기 때문에 개발자가 굳이 같은 상태 코드를 `ErrorCode`에 일일히 적지 않아도 좋을 것 같다는 느낌이 들었습니다! 

단점으로는 한 에러 카테고리 내에서 특정 코드가 상이한 상태 코드를 가지고 있는 경우에는 유연성이 떨어질 수 있는 점이지만 이 부분에 대해서는 `BusinessException`의 하위 클래스를 적절한 수로 세분화 시키는 방법 등으로 상쇄시킬 수 있을 것 같다는 생각이 들었습니다! 하지만 `BusinessException`의 하위 클래스의 수가 너무 많아지면 `ErrorCode`의 의미가 퇴색되므로 적절한 조율이 필요할 것 같아요.

우선 `HttpStatus`를 제외했다가 또 다시 세부적인 상태 코드가 필요해진다면 포함시키는 것도 좋을 것 같습니다! 😄

thornscript reviewed (commented):
저같은 경우는 이런 코드들을 처음 접했을 때 약간 관용구 같이 사용되는 코드들이 많아서 사실 처음에는 '왜 그렇게 작성해야 할까?'라는 의문이 들기도 했었는데, 전체적으로 잘 작성해주신 것 같고 다수의 타 프레임워크 경험이 있으시다보니까 자바나 스프링에 대해서도 금방 적응하실 것 같은 느낌이 들었습니다!

수고 많으셨습니다 🥹

thornscript commented on src/main/java/com/growup/pms/common/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({AuthenticationException.class})
+    protected ResponseEntity<ErrorDto> handleAuthenticationException(AuthenticationException authenticationException) {
+        return ResponseEntity
+                .status(authenticationException.getErrorCode().getHttpStatus())
+                .body(authenticationException.getErrorDto());
+    }
+
+    @ExceptionHandler({AuthorizationException.class})
+    protected ResponseEntity<ErrorDto> handleAuthorizationException(AuthorizationException authorizationException) {
+        return ResponseEntity
+                .status(authorizationException.getErrorCode().getHttpStatus())
+                .body(authorizationException.getErrorDto());
+    }
+
+    @ExceptionHandler({EntityNotFoundException.class})
+    protected ResponseEntity<ErrorDto> handleEntityNotFoundException(EntityNotFoundException entityNotFoundException) {
+        return ResponseEntity
+                .status(entityNotFoundException.getErrorCode().getHttpStatus())
+                .body(entityNotFoundException.getErrorDto());
+    }
```
특정 예외에 대해서 상이한 처리를 해주어야 할 때 이렇게 정의해주어도 괜찮을 것 같고, 혹시 공통되는 부분이라면 아래와 같이 부모 클래스인 `BusinessException.class`를 잡아도 괜찮을 것 같아요!

```java
@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(BusinessException.class)
    protected ResponseEntity<ErrorDto> handleBusinessException(BusinessException e) {
        var response = ErrorDto.of(e.getErrorCode());
        log.error("Unexpected error has occurred: {}", e.getMessage());
        return new ResponseEntity<>(response, response.getHttpStatus());
    }

    // 미처 잡지 못한 스프링 내부 예외를 잡기 위한 예외 핸들러 (비정상 흐름)
    @ExceptionHandler(Exception.class)
    protected ResponseEntity<ErrorDto> handleException(Exception e) {
        var response = ErrorDto.of(ErrorCode.INTERNAL_SERVER_ERROR);
        log.error(e.getMessage(), e);
        return new ResponseEntity<>(response, response.getHttpStatus());
    }
}
```

thornscript commented on src/test/java/com/growup/pms/common/exception/handler/GlobalExceptionHandlerTest.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.context.annotation.Import;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@Import(GlobalExceptionHandler.class)
+@ExtendWith(SpringExtension.class)
+class GlobalExceptionHandlerTest {
+
+    @Test
+    @DisplayName("EntityNotFoundException 테스트를 한다.")
+    void entityNotFoundException() {
+        // given & when & then
+        Assertions.assertThrows(EntityNotFoundException.class, () -> {
+            throw new EntityNotFoundException(ErrorCode.ENTITY_NOT_FOUND);
+        });
+    }
+
+    @Test
+    @DisplayName("AuthenticationException 테스트를 한다.")
+    void authenticationException() {
+        // given & when & then
+        Assertions.assertThrows(AuthenticationException.class, () -> {
+            throw new AuthenticationException(ErrorCode.AUTH_AUTHENTICATION_ERROR);
+        });
+    }
+
+    @Test
+    @DisplayName("AuthorizationException 테스트를 한다.")
+    void authorizationException() {
+        // given & when & then
+        Assertions.assertThrows(AuthorizationException.class, () -> {
+            throw new AuthorizationException(ErrorCode.AUTHZ_ACCESS_DENIED);
+        });
+    }
+}
```
`@Import` 요 부분에 대해서는 `GlobalExceptionHandler`가 설정 클래스(configuration classes)가 아니다 보니까 'Could not detect default configuration classes'라는 메시지를 로그에서 보셨을 것 같아요! 아무래도 `GlobalExceptionHandler`의 기능이 AOP랑 엮여있다 보니까 임의의 컨트롤러를 하나 만들고 테스트를 할 수 있을 것 같기는 합니다!

```java
@WebMvcTest(GlobalExceptionHandler.class)
public class GlobalExceptionHandlerTest {

    @Autowired
    private MockMvc mockMvc;

    @TestConfiguration
    public static class TestConfig {

        @RestController
        @RequestMapping("/test")
        public static class TestController {

            @GetMapping("/auth")
            public String test() {
                throw new AuthenticationException(ErrorCode.AUTH_AUTHENTICATION_ERROR);
            }
        }
    }

    @Test
    @WithUserDetails
    public void testHandleAuthenticationException() throws Exception {
        mockMvc.perform(get("/test/auth"))
                .andExpect(status().isUnauthorized())
                .andExpect(jsonPath("$.code").value(ErrorCode.AUTH_AUTHENTICATION_ERROR.getCode()))
                .andExpect(jsonPath("$.message").value(ErrorCode.AUTH_AUTHENTICATION_ERROR.getMessage()));
    }
}
```

KimJayhyun replied:
경험이 많으시니 코드의 퀄리티가 확 다르네요..!

`@TestConfiguration`는 처음 봤어요..!
Test를 위해 만들었던 Controller를 대체할 수가 있네요..
Controller로 테스트하고 싶은 제 고민을 해결하는 방법이네요!

내용들을 정리해가면서 해야 `@Import`와 같이 또 실수하는 일이 없을 텐데 말이죠... 

도현님의 리뷰에 공부가 너무 잘 되가고 있습니다. 
정리를 하는 시간을 한 번 가져서 실수가 없도록 노력할게요!

thornscript replied:
@KimJayhyun 사실 저도 동료들의 피드백을 받고 무한 성장한 경험이 많기도 하고, 이렇게 제가 알고 있는 지식을 공유하고 동료들과의 논의로 재점검하면서 또 크게 성장하는 부분이 있는 것 같아요. 여담으로 이런 부분 하나하나가 면접 질문으로 들어올 수 있기도 하...

아니면 테스트 환경에서만 사용하는 `test` 프로필을 만들고 `@ActiveProfiles` 애노테이션으로 테스트 환경에서만 빈으로 등록될 수 있게끔 할 수도 있을 것 같아요! 자신이 선호하시는 방법을 찾아서 작성하시면 좋을 것 같습니다 👍

thornscript commented on src/main/java/com/growup/pms/common/exception/exceptions/ErrorCode.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+public enum ErrorCode {
+
+    ENTITY_NOT_FOUND(HttpStatus.BAD_REQUEST, "ET_001", "Entity error: ENTITY가 없습니다."),
+
+    AUTH_AUTHENTICATION_ERROR(HttpStatus.UNAUTHORIZED, "AT_001","Authentication error : 인증을 실패했습니다."),
+
+    AUTHZ_ACCESS_DENIED(HttpStatus.FORBIDDEN, "ATZ_001","Authorization error : 접근권한이 없습니다.");
```
이 부분은 사실 제가 리뷰하기는 애매해서 GPT의 의견을 물어봤는데 아래와 같은 의견을 들을 수 있었어요 🥹 

> 1. **메시지 포맷 일관성:** 메시지에 있는 공백이나 콜론(:) 사용에서 일관성을 유지하는 것이 좋습니다. AUTH_AUTHENTICATION_ERROR의 경우 "Authentication error : 인증을 실패했습니다."와 같이 콜론 앞에 공백이 있는 반면, 다른 메시지에는 없습니다.
> 2. **메시지 내용의 명확성:** 한국어와 영어가 혼합된 메시지를 사용할 때는 명확하고 일관되게 작성하는 것이 중요합니다. 예를 들어, 모든 메시지를 영어로 작성하거나 모든 메시지를 한국어로 작성하는 것이 더 나을 수 있습니다.

<p align="center"><img src="https://github.com/GU-99/grow-up-pms/assets/8318212/3b36d3d8-e13a-46f1-b0a8-c71de26a1324"></p>

KimJayhyun replied:
제가 사소한 부분들을 자꾸 놓치네요.. ㅠㅠ
`메세지 포멧 일관성`을 지키도록 다시 작성하도록 할게요

`메세지 내용의 명확성`에 대한 제 생각은 아래에 적어볼게요
- 우리가 ErrorCode를 사용하는 이유는 개발자와 사용자 모두에게 알아보기 쉬운 메세지를 전하기 위함이다...!
- 알아보기 쉽기 위해서는 **한글을 사용하는 것**이 좋다고 생각해요.
- 하지만 `Entity` 같은 특정 명사들은 영어가 알아보고 인지하기 쉽지 않을까요??
- 또 에러 코드에 대해 대분류(`Entity error : `와 같이 `:`의 앞 부분)를 설정해서 빠른 이해를 도와주는 것도 좋다고 생각을 했어요

도현님은 어떻게 생각하시나요..!?

KimJayhyun replied:
아휴 이런 작은 기능을 넣는데도 많은 생각을 하게 되네요..!

thornscript replied:
@KimJayhyun 이건 정말 개발자마다 의견이 갈릴 것 같기는 한데, 사실 HTTP 상태 코드의 가짓수에는 한계가 있기도 하고 상황에 따라서 추상적일 때도 있는데다가 축약된 에러 카테고리(예: ATZ)를 보고 클라이언트 측에서는 API 명세서를 보는 게 아닌 이상은 쉽게 인지할 수 없다는 문제점이 있는 것 같아요! 그래서 재현님이 제안해주신 방법이 도움이 될 것 같은데, 각 에러 카테고리에 대해서 공통된 접두사를 가진다면 상수처럼 한 곳에서 관리할 수 있다면 좋을 것 같다는 생각이 들었습니다 🐱

KimJayhyun committed:
Feat: PR conversation을 해결했어요

Commit:
Style: 컨벤션에 맞도록 쉼표 앞뒤로 띄어쓰기 추가

Commit:
Refactor: BusinessException에서 ErrorResponse 필드 제거

BusinessException 클래스에서 ErrorResponse 필드를 제거하였습니다.
ErrorResponse는 ErrorCode로부터 생성할 수 있으므로 예외 클래스에 직접 포함할 필요가 없습니다.
이로 인해 클래스의 단순성과 책임이 분리되어 코드의 명확성을 높일 수 있습니다.

Commit:
Style: 불필요한 중괄호 제거

Commit:
Refactor: 불필요한 주석 제거

handleException 메서드의 주석을 제거하였습니다.
코드 자체가 충분히 명확하므로 주석이 불필요합니다.

Commit:
Refactor: 추적의 용이함을 위해 예외 핸들러에 로그 메시지를 같이 기록함

Commit:
Feat: 가독성을 위해 언더바를 띄어쓰기로 전환하도록 애노테이션을 추가함

thornscript reviewed (approved):
수고하셨습니다! 빠르게 머지하겠습니다 👍
